src/interpreter.jl

```diff
@@ -107,38 +107,35 @@ function InferredSource(state::InferenceState)
         exct)
 end
 
-function cthulhu_finish(@specialize(finishfunc), state::InferenceState, interp::CthulhuInterpreter, cycleid::Int)
-    res = @invoke finishfunc(state::InferenceState, interp::AbstractInterpreter, cycleid::Int)
+function CC.finishinfer!(state::InferenceState, interp::CthulhuInterpreter, cycleid::Int)
+    res = @invoke CC.finishinfer!(state::InferenceState, interp::AbstractInterpreter, cycleid::Int)
     key = get_inference_key(state)
     if key !== nothing
         interp.unopt[key] = InferredSource(state)
     end
     return res
 end
 
-function create_cthulhu_source(result::InferenceResult, effects::Effects)
-    isa(result.src, OptimizationState) || return result.src
-    opt = result.src
+function CC.transform_result_for_local_cache(interp::CthulhuInterpreter, result::InferenceResult)
+    src = @invoke CC.transform_result_for_local_cache(interp::AbstractInterpreter, result::InferenceResult)
+    isa(src, OptimizationState) || return src
+    opt = src
     @static if VERSION ≥ v"1.13-"
         optresult = opt.optresult::CC.OptimizationResult
         optresult.simplified || CC.simplify_ir!(optresult)
-        opt.src.inlining_cost = CC.compute_inlining_cost(opt.inlining.interp, result, optresult)
         ir = CC.compact!(copy(optresult.ir))
     else
         # get the (theoretically) same effect as the jl_compress_ir -> jl_uncompress_ir -> inflate_ir round-trip
         ir = CC.compact!(CC.cfg_simplify!(CC.copy(opt.ir::IRCode)))
     end
-    return OptimizedSource(ir, opt.src, opt.src.inlineable, effects)
+    return OptimizedSource(ir, opt.src, opt.src.inlineable, result.ipo_effects)
 end
 
-function set_cthulhu_source!(result::InferenceResult)
-    result.src = create_cthulhu_source(result, result.ipo_effects)
-end
-
-CC.finishinfer!(state::InferenceState, interp::CthulhuInterpreter, cycleid::Int) = cthulhu_finish(CC.finishinfer!, state, interp, cycleid)
-function CC.finish!(interp::CthulhuInterpreter, caller::InferenceState, validation_world::UInt, time_before::UInt64)
-    set_cthulhu_source!(caller.result)
-    return @invoke CC.finish!(interp::AbstractInterpreter, caller::InferenceState, validation_world::UInt, time_before::UInt64)
+function CC.transform_result_for_cache(interp::CthulhuInterpreter, result::InferenceResult, edges::Core.SimpleVector)
+    if isa(result.src, CodeInfo)
+        result.src.edges = edges
+    end
+    result.src = CC.transform_result_for_local_cache(interp, result)
```

Comment:
@aviatesk do you know if it is safe to discard edges for `OptimizationState`? My understanding is that edges anyways come from `opt.inlining.edges`, so the structure already has them, but before that we used to take the `OptimizationState` -> `CodeInfo` path and attach edges to the resulting `CodeInfo`.